Adding New Modules

jbadger95 opened the pull request:
Adding a few modules that I used to support some other changes.

- bitvisit -- Stores and manipulates array of visitation flags (optionally bitpacks flags) that is convenient for reducing across MPI ranks
- index_map -- Maps a collection of array items to their location (index) in the array
- sorts -- A collection of sorting routines used throughout the code

stefanhenneking commented:
Are there any sorting routines that you copy-pasted from elsewhere in this new module and which can now be removed elsewhere?

stefanhenneking commented on trunk/src/modules/sorts.F90:
```diff
@@ -0,0 +1,418 @@
+!! @param[in]    Last      - last index of current partition (needs to be N on initial call)
+   recursive subroutine qsort_duplet(Iel_array,Residuals,N,First,Last)
+!
+      implicit none
```
```suggestion
```

stefanhenneking commented on trunk/src/modules/sorts.F90:
```diff
@@ -0,0 +1,418 @@
+!! @param[in]    last  - last index of current partition (N on initial call)
+   recursive subroutine qsort_triplet(ia,ja,xa,n,first,last)
+!
+      implicit none
```
```suggestion
```

stefanhenneking commented on trunk/src/modules/sorts.F90:
```diff
@@ -0,0 +1,418 @@
+!! @param[in]    last  - last index of current partition (N on initial call)
+   recursive subroutine partition_triplet_omp(ip,jp,n,k,ia,ja,xa)
+!
+      implicit none
```
```suggestion
```

stefanhenneking commented on trunk/src/modules/sorts.F90:
```diff
@@ -0,0 +1,418 @@
+!> Purpose - Threaded version of qsort_triplet
+recursive subroutine qsort_triplet_omp(n,ia,ja,xa)
+!
+   implicit none
```
```suggestion
```

stefanhenneking commented on trunk/src/modules/sorts.F90:
```diff
@@ -0,0 +1,418 @@
+!> Purpose - sorts duplet (ia,val) by index
+   recursive subroutine qsort_double(ia,val,n,first, last)
+!
+      implicit none
```
```suggestion
```

jbadger95 commented:
I tried to copy all the sorting routines (don't know if I got all of them). I have not replaced them in the code yet though; to do this routines using sorts just need to use the `sorts` module and the routines need to be deleted; I can do this but I wasn't sure if you thought it would be more clear to have some of these functions defined where they are used (like in the sparse matrix routines). The main reason for committing this was that my `index_map` module needs one of the routines and I already had them all modularized. `index_map` is used in some uncommited changes for pre-computing element nodes (we decided we may not add that, I still may add the module after testing though; it doesn't have to be used if not wanted). It's also used heavily in my MG solver

stefanhenneking commented:
I like the idea of having a sort module that's used consistently in the code and remove sorting functions elsewhere. But we can wait with the removal of other sorting functions until having merged #75 which removes some of them also to avoid the conflicts.

stefanhenneking commented on trunk/src/modules/sorts.F90:
```diff
@@ -0,0 +1,418 @@
+
+
+!> Purpose - partitions triplet array around a pivot
+!! @param[inout] ia    - integer array (row indices)
+!! @param[inout] ja    - integer array (colun indices)
+!! @param[inout] xa    - array of values
+!! @param[out]   k     - pivot
+!! @param[in]    n     - list of integers to sort
+!! @param[in]    first - first index of current partition (1 on initial call)
+!! @param[in]    last  - last index of current partition (N on initial call)
```
update params

stefanhenneking commented:
@jbadger95 can you check on the comment/suggestions. then, it's ready to be merged imo.

jbadger95 commented on trunk/src/modules/bitvisit.F90:
```diff
@@ -0,0 +1,213 @@
+      if (bitvisit_packed) then
+!     ...get word number for entry
+         iword = (i - 1)/bpw + 1
+!$omp critical
+         if (val .eq. 1) then
+            visitation(iword) = ibset(visitation(iword),mod(i-1,bpw))
+         else
+            visitation(iword) = ibclr(visitation(iword),mod(i-1,bpw))
+         endif
+!$omp end critical
```
visitation is the array that stores whether something has been visited (it is sized when the initialization routine is called). It's frequently of size NRNODS so no, threadprivate is not an option.

stefanhenneking replied:
ok, maybe we can yet think of another way to "visit" more efficiently in the future (of course, if `NODES` were distributed, they issue becomes more or less irrelevant in terms of performance).

jbadger95 commented:
Ok, I addressed most of the comments. Removing redundant `implicit none`s is a minor stylistic difference (I like having it in the functions for clarity) so I think we can let these slide.

stefanhenneking commented:
Sounds good, merge when you're ready.
